ASoC: SOF: Intel: hda-dai: fix suspend and pause/release cases

plbossart opened the pull request:
This is an attempt to finish what we started for deep-buffer support and never completed.

Compile-tested only, the last two patches are the most critical ones.

ranj063 commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -251,8 +254,10 @@ static int hda_link_dma_trigger(struct snd_pcm_substream *substream, int cmd)
 		}
-
+		snd_soc_dai_set_dma_data(cpu_dai, substream, NULL);
+		snd_hdac_ext_stream_release(hext_stream, HDAC_EXT_STREAM_TYPE_LINK);
```
@plbossart is this working without errors for normal playback/capture stop? I think you might need to add if(hext_stream->link_prepared) return 0; in hda_link_dma_hw_free() no?

plbossart replied:
I reverted this change for now, let's try with the code split first to see if I screwed up.

plbossart commented:
2c1b27867c460f761f39e95cd49f93b3ac4f4564 is the first bad commit
commit 2c1b27867c460f761f39e95cd49f93b3ac4f4564
Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Date:   Mon Aug 30 22:33:18 2021 -0500

    ASoC: SOF: Intel: hda-dai: split link DMA and dai operations
    
    The link DMA state management is handled completely on the host side,
    while the DAI operations require an IPC. Split the first part in
    dedicated helpers.
    
    Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

 sound/soc/sof/intel/hda-dai.c | 222 ++++++++++++++++++++++++++----------------
 1 file changed, 138 insertions(+), 84 deletions(-)

plbossart commented:
@ranj063 if I add this diff things work again

````diff
diff --git a/sound/soc/sof/intel/hda-dai.c b/sound/soc/sof/intel/hda-dai.c
index facdd45bdd29..68a5af46a084 100644
--- a/sound/soc/sof/intel/hda-dai.c
+++ b/sound/soc/sof/intel/hda-dai.c
@@ -214,9 +214,6 @@ static int hda_link_dma_prepare(struct snd_pcm_substream *substream)
        struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
        int stream = substream->stream;
 
-       if (hext_stream->link_prepared)
-               return 0;
-
        dev_dbg(sdev->dev, "%s: prepare stream dir %d\n", __func__, substream->stream);
 
        return hda_link_dma_hw_params(substream, &rtd->dpcm[stream].hw_params);
@@ -429,11 +426,16 @@ static int hda_dai_config_pause_push_ipc(struct snd_soc_dapm_widget *w)
 static int hda_dai_prepare(struct snd_pcm_substream *substream,
                           struct snd_soc_dai *dai)
 {
-       struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+       struct hdac_ext_stream *hext_stream =
+                               snd_soc_dai_get_dma_data(dai, substream);
        struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
+       struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
        int stream = substream->stream;
        int ret;
 
+       if (hext_stream->link_prepared)
+               return 0;
+
        dev_dbg(sdev->dev, "%s: prepare stream dir %d\n", __func__, substream->stream);
 
        ret = hda_link_dma_prepare(substream);
````

So it  looks like I was sending the same IPC twice, any idea why this could cause the DSP firmware to panic?

ranj063 commented:
> So it looks like I was sending the same IPC twice, any idea why this could cause the DSP firmware to panic?

hmm but @plbossart there is no IPC in hda_dai_prepare() no?

plbossart commented:
> > So it looks like I was sending the same IPC twice, any idea why this could cause the DSP firmware to panic?
> 
> hmm but @plbossart there is no IPC in hda_dai_prepare() no?

it's not shown by the diff but here's the code:

````c

static int hda_dai_hw_params_update(struct snd_pcm_substream *substream,
				    struct snd_pcm_hw_params *params,
				    struct snd_soc_dai *dai)
{
	struct hdac_ext_stream *hext_stream;
	struct snd_soc_dapm_widget *w;
	int stream_tag;

	hext_stream = snd_soc_dai_get_dma_data(dai, substream);
	if (!hext_stream)
		return -EINVAL;

	stream_tag = hdac_stream(hext_stream)->stream_tag;

	w = snd_soc_dai_get_widget(dai, substream->stream);

	/* set up the DAI widget and send the DAI_CONFIG with the new tag */
	return hda_dai_widget_update(dai->dev, w, stream_tag - 1, true);
}

static int hda_dai_prepare(struct snd_pcm_substream *substream,
			   struct snd_soc_dai *dai)
{
	struct hdac_ext_stream *hext_stream =
				snd_soc_dai_get_dma_data(dai, substream);
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(dai->component);
	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
	int stream = substream->stream;
	int ret;

	if (hext_stream->link_prepared)
		return 0;

	dev_dbg(sdev->dev, "%s: prepare stream dir %d\n", __func__, substream->stream);

	ret = hda_link_dma_prepare(substream);
	if (ret < 0)
		return ret;

	return hda_dai_hw_params_update(substream, &rtd->dpcm[stream].hw_params, dai);
}

````

so what I was doing is sending an update twice. This seems to be fatal for the DSP...

marc-hb commented:
https://sof-ci.01.org/linuxpr/PR3331/build6907/devicetest/ had only one failure:
```
systemctl status gpm
● gpm.service - LSB: gpm sysv init script
     Loaded: loaded (/etc/init.d/gpm; generated)
     Active: failed (Result: exit-code) since Wed 2021-12-15 20:31:07 UTC; 44min ago
       Docs: man:systemd-sysv-generator(8)
    Process: 505 ExecStart=/etc/init.d/gpm start (code=exited, status=2)

Dec 15 20:31:07 sh-adlp-rvp-sdw-01 systemd[1]: Starting LSB: gpm sysv init script...
Dec 15 20:31:07 sh-adlp-rvp-sdw-01 gpm[505]:  * Starting mouse interface server gpm
Dec 15 20:31:07 sh-adlp-rvp-sdw-01 gpm[505]: /etc/init.d/gpm: 66: cannot open /dev/input/mice: No such file
Dec 15 20:31:07 sh-adlp-rvp-sdw-01 systemd[1]: gpm.service: Control process exited, code=exited, status=2/INVALIDARGUMENT
Dec 15 20:31:07 sh-adlp-rvp-sdw-01 systemd[1]: gpm.service: Failed with result 'exit-code'.
Dec 15 20:31:07 sh-adlp-rvp-sdw-01 systemd[1]: Failed to start LSB: gpm sysv init script.
```
Fixed by `apt remove gpm`. No idea why this system had it.

ranj063 commented:
@plbossart there are conflicts in this PR mainly in hda-dai.c because of my recent changes

plbossart commented:
replaced by https://github.com/thesofproject/linux/pull/3515
